src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2230,16 +2227,32 @@ def unlock(self, job, id, options):
 
             try:
                 self.middleware.call_sync(
-                    'zfs.dataset.load_key', name, {'key': datasets[name]['key'], 'mount': True, 'recursive': True}
+                    'zfs.dataset.load_key', name, {'key': datasets[name]['key'], 'mount': False}
                 )
             except CallError as e:
                 failed[name]['error'] = 'Invalid Key' if 'incorrect key provided' in str(e).lower() else str(e)
             else:
-                unlocked.append(name)
+                # Before we mount the dataset in question, we should ensure that the path where it will be mounted
+                # is not already being used by some other service/share. In this case, we should simply rename the
+                # directory where it will be mounted
+
+                mount_path = os.path.join('/mnt', name)
+                if os.path.exists(mount_path):
+                    if not os.path.isdir(mount_path) or os.listdir(mount_path):
+                        # rename please
+                        shutil.move(mount_path, f'{mount_path}-{str(uuid.uuid4())[:4]}-{datetime.now().isoformat()}')
 
-        if options['toggle_attachments'] and not failed:
-            j = self.middleware.call_sync('pool.dataset.toggle_attachments', self.__attachments_path(dataset))
-            j.wait_sync()
+                try:
+                    self.middleware.call_sync('zfs.dataset.mount', name, {'recursive': True})
+                except CallError as e:
+                    failed[name]['error'] = f'Failed to mount dataset: {e}'
+                else:
+                    unlocked.append(name)
+
+        if options['restart_attachments']:
```

Comment:
If `restart_attachments` is false, then we'll be left with system in inconsistent state (some shares are present in the database and are not present in config). What is the reason not to restart attachments? All our share services can reload their config w/o interrupting connections, can't they?

Reply:
@themylogin so we do not want to restart attachments on boot when we are unlocking datasets which means we have to have a knob to configure this. We can instead remove the option in the UI and we already default to true, so it will restart related attachments when user unlocks it. @william-gr also pointed out that we shouldn't be changing the api and it's not really necessary. Please let me know your thoughts on this one and I can make changes accordingly.

Reply:
Thank you for explanation, I agree with it being a private option for internal/advanced use.